camply/notifications/slack.py

```diff
@@ -0,0 +1,125 @@
+                "proper environment variable: `SLACK_WEBHOOK`."
+            )
+            logger.error(warning_message)
+            raise EnvironmentError(warning_message)
+
+    def __repr__(self):
+        """
+        String Representation
+        """
+        return "<SlackNotifications>"
+
+    @staticmethod
+    def send_message(message: str, **kwargs) -> requests.Response:
+        """
+        Send a message via Slack - if environment variables are configured.
+
+        Parameters
+        ----------
+        message: str
+        blocks: List
+
+        Returns
+        -------
+        requests.Response
+        """
+        slack_headers = {"Content-Type": "application/json"}
+
+        message_blocks = kwargs.pop("blocks", [])
+        message_json = {
+            "text": message,
+        }
+        if message_blocks:
+            message_json = {
+                "blocks": message_blocks,
+            }
+        logger.debug(message_json)
+        response = requests.post(
+            url=SlackConfig.SLACK_WEBHOOK,
+            headers=slack_headers,
+            json=message_json,
+        )
+        try:
+            response.raise_for_status()
+        except requests.HTTPError as he:
+            logger.warning(
+                "Notifications weren't able to be sent to Slack. "
+                "Your configuration might be incorrect."
+            )
+            raise ConnectionError(response.text) from he
+        return response
+
+    @classmethod
+    def send_campsites(cls, campsites: List[AvailableCampsite], **kwargs):
+        """
+        Send a message with a campsite object
+
+        Parameters
+        ----------
+        campsites: AvailableCampsite
+        """
+        for campsite in campsites:
+            message_title, formatted_dict = cls.format_standard_campsites(
+                campsite=campsite,
+            )
+            fields = []
+            for key, value in formatted_dict.items():
+                fields.append({
+                    "type": "mrkdwn",
+                    "text": f"*{key}*",
+                })
+                if key in ["Permitted Equipment", "Booking Link"]:
+                    data_type = "mrkdwn"
+                else:
+                    data_type = "plain_text"
+                fields.append({
+                    "type": data_type,
+                    "text": str(value),
+                })
+
+            blocks = []
+            blocks.append({
+                "type": "header",
+                "text": {
+                    "type": "plain_text",
+                    "text": message_title,
+                },
+            })
+            # Slack only allows 10 fields (k+v) per section
+            for chunk in range(0, math.ceil(len(fields) / 10)):
+                blocks.append({
+                    "type": "section",
+                    "fields": fields[chunk * 10:chunk * 10 + 10],
+                })
```

Comment:
Everything looks great - what an awesome implementation. This is really my only question. what exactly is going on here?

Reply:
`fields` contains a list of dicts, each corresponding to keys and values from `formatted_dict` in alternating order, as they would be passed to "section" blocks in the slack message. We have to chunk these as Slack [only allows 10 fields per section](https://api.slack.com/reference/block-kit/blocks#section).

I just realized that `range()` can take an increment arg so I can make this a bit cleaner.

Reply:
Oh, very cool. This all makes sense. I love the use of blocks instead of just a plain text message too - this is an awesome new feature. These new Slack updates might just make me changeover from Pushover